File: Programmers/올바른 괄호/올바른 괄호.py
def solution(s):
    answer = True
    stack = []
    for i in range(len(s)):
        if s[i] == '(':
            stack.append(s[i])
        else:
            if len(stack) == 0:
                return False
            else:
                stack.pop()
    if len(stack) > 0:
        return False
    else:
        return True

File: Java/Etc/올바른 괄호.java
import java.util.*;

public class Main {

    public String solution (String data) {
        String answer = "YES";
        Stack<Character> stack = new Stack<>();

        for (char x : data.toCharArray()) {
            if (x == '(') {
                stack.push(x);
            } else {
                if (stack.isEmpty()) return "NO";
                else{
                    if (stack.get(stack.size()-1) == '(') {
                        stack.pop();
                    } else return "NO";
                }
            }
        }
        if (!stack.isEmpty()) return "NO";
        return answer;
    }

    public static void main(String[] args) {
        Main T = new Main();
        Scanner scanner = new Scanner(System.in);
        String data = scanner.next();

        System.out.println(T.solution(data));
    }
}

